Editorial workflow dependency tracking

Benaiah opened the pull request:
<!--
Thanks for submitting a pull request!

Please make sure you've read and understood our contributing guidelines;
https://github.com/netlify/netlify-cms/blob/master/CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx", where #xxxx is the issue number.

Please provide enough information so that others can review your pull request.
The first three fields are mandatory:
-->

**- Summary**

Dependency tracking for unpublished entries: closes #192. 

**- Test plan**

- Existing test suite passes.
- State actions manually tested by dispatching appropriate actions. 
- Dependency resolution and publishing of multiple entries manually tested on `benaiah/netlify-cms-test-repo`.
- pre-commit hooks skipped since they complain about code already in master.

**- Description for the changelog**

- **Add dependency tracking and multiple-entry publishing.**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://cloud.githubusercontent.com/assets/1425133/23242412/7ddd55e0-f92d-11e6-953f-40f4412c4f6e.png)

Benaiah commented on src/reducers/editorialWorkflow.js:
```diff
@@ -76,6 +77,15 @@ const unpublishedEntries = (state = null, action) => {
       return state.deleteIn(['entities', `${ action.payload.collection }.${ action.payload.slug }`]);
 
+    case UNPUBLISHED_ENTRY_REGISTER_DEPENDENCY:
+      return state.setIn([
```
This is currently storing the dependencies in `entry.data.dependencies`, where it is then grabbed in `persistUnpublishedEntry`, added to `options`, and stored in metadata in `editorialWorkflowGit`. Should this be saving directly to `entry.metaData` instead?

calavera commented on example/config.yml:
```diff
@@ -1,58 +1,29 @@
 backend:
-  name: test-repo
```
please, do not push changes to this file

Benaiah replied:
This is temporary, it'll be rebased out before this leaves WIP.

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -180,6 +187,34 @@ function unpublishedEntryPublishError(collection, slug, transactionID) {
+  };
+}
+
+function unpublishedEntryDependenciesRequest (collection, slug) {
```
please, remove space between method name and parameters.

Benaiah replied:
Will do (ditto to those below).

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -180,6 +187,34 @@ function unpublishedEntryPublishError(collection, slug, transactionID) {
+  };
+}
+
+function unpublishedEntryDependenciesSuccess (collection, slug, dependencies) {
```
please, remove space between method name and parameters.

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -180,6 +187,34 @@ function unpublishedEntryPublishError(collection, slug, transactionID) {
+  };
+}
+
+function unpublishedEntryDependenciesError (collection, slug, error) {
```
please, remove space between method name and parameters.

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+
+export const registerUnpublishedEntryDependency = unpublishedEntryRegisterDependency;
+
+export function getUnpublishedEntryDependencies (collection, slug) {
```
please, remove space between method name and parameters.

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+      "dependencies",
+    ];
+
+    const getEventualDependencies = (paths, loadedDeps, state) =>
```
why is this function defined inside the parent function? please, extract.

Benaiah replied:
This function is only useful within `getUnpublishedEntryDependencies` - I didn't want it exposed to the outer scope since it's very specific to the functionality of `traverse`. I'm fine with extracting it, that was just the reasoning behind nesting it with `traverse` (which can't be hoisted).

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+           .then(() => path);
+       });
+
+    const traverse = (collectedDeps, path) => {
```
why is this function defined inside the parent function? please, extract.

Benaiah replied:
This function closes over `getState`, which is why it was defined here. I can hoist it and make `getState` an argument, though.

calavera commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+    };
+
+    // Begin traversal
+    return traverse(new Immutable.List(), `${ collection }.${ slug }`)
```
If you use a Map rather than a List, you'll avoid the N lookup per iteration in the `traverse` call.

Benaiah replied:
Good point, thanks!

Benaiah commented:
@calavera I've addressed your comments on `editorialWorkflow.js` - thanks for the review!

biilmann commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+  const state = getState();
+
+  if (collectedDeps.get(path) === true) {
+    return new Promise(resolve => resolve(collectedDeps));
```
`return Promise.resolve(collectedDeps)` is cleaner

biilmann commented on src/actions/editorialWorkflow.js:
```diff
@@ -291,3 +326,66 @@ export function publishUnpublishedEntry(collection, slug) {
+  if (!state.editorialWorkflow.hasIn(depsPath) ||
+    state.editorialWorkflow.getIn(depsPath).size === 0) {
+    // This path has no dependencies
+    return new Promise(resolve => resolve(newDeps));
```
`return Promise.resolve(newDeps)`

biilmann commented on src/reducers/entryDraft.js:
```diff
@@ -73,6 +74,10 @@ const entryDraftReducer = (state = Map(), action) => {
       return state.update('mediaFiles', list => list.filterNot(path => path === action.payload));
 
+    case UNPUBLISHED_ENTRY_REGISTER_DEPENDENCY:
+      return state.setIn(['entry', 'data', 'dependencies', action.payload.field],
```
The entry data is what gets persisted in the repository - we don't want to add "dependencies" to the frontmatter of the entry and the name could easily class with a field.

Either we should store them in something like `entry.dependencies` or do like the assets and have `dependencies` in the state for entryDraft.

Benaiah replied:
Ah, thanks for clarifying that. We definitely don't want this clashing with frontmatter.

biilmann commented on src/components/Widgets/ControlHOC.js:
```diff
@@ -80,14 +81,25 @@ class ControlHOC extends Component {
       onChange,
       onAddAsset,
       onRemoveAsset,
+      registerDependency,
```
seems like `onAddDependency` and `onRemoveDependency` would be more in line with the other callback functions in terms of naming (and it should be possible to remove a dependency as well as adding one...)

Benaiah replied:
Yeah, that's a good point re: naming. See my comment below for more on the removal and updating.

Benaiah commented:
@biilmann *updating* a dependency already works (in the case of a 1-to-1 field->dependency relationship, there isn't any functionality to handle a 1-to-many field->dependencies relationship). *Removing* a dependency is unimplemented - perhaps this should just be triggered whenever a relation field changes (other than selecting a suggestion)?

Benaiah commented:
@biilmann I've pushed changes that address your comments. I've moved the dependency storage to `entry.dependencies`, renamed the `registerDependency` prop to `onAddDependency`, added an `onRemoveDependency` prop (this is currently a noop), and changed the immediately-resolving promises to use `Promise.resolve`.

I'm working on hooking this into the `publishUnpublishedEntry` action, and I noticed something - if I do the simplest implementation, the traversal doesn't work as the current entry is optimistically removed from the `editorialWorkflow` store on the publish request. I could either grab the dependencies from the entry object itself, which would be a bit messier as `traverse` would be implemented two different ways (once for the root entry, once for the rest); or I could only fire the publish request *after* the dependency check. Thoughts?

calavera commented:
> or I could only fire the publish request after the dependency check. Thoughts?

I think firing the publish request when the we're sure everything is correct might be more expected. @biilmann wdyt?

Benaiah commented:
@calavera @biilmann I've pushed an updated version with all the functionality, including dependency removal. It currently uses `window.confirm` for UI - I think building out a custom modal interface should wait until the UI stuff has settled down a bit, and the search issue I've mentioned previously (unpublished dependencies don't show up in relation widget suggestions) means you can't register dependencies yet from the UI anyway.

Benaiah commented:
@calavera @biilmann I've clarified `traverse` a bit more by pulling the promise array reduction into its own function, `reducePromises`. The function seems pretty out-of-place in `src/actions/editorialWorkflow.js` though - what do you think about moving it somewhere, perhaps `lib/promiseHelper`?

Benaiah commented:
@calavera @biilmann I've removed the temporary config.yml commit that I was using for testing.

calavera commented:
LGTM, but unfortunately, there are some conflicts.

Benaiah commented:
@calavera I'll get it rebased again.

Benaiah commented:
Blech, tried to git in the morning and ended up accidentally squashing two commits. It looks like all the changes are still there, but the changes ended up in two commits instead of the four I had before.

Benaiah commented:
@calavera, @biilmann OK, got that fixed. It's now rebased onto current master with the correct history (thanks, `git reflog`).

Benaiah commented:
I'm closing this PR for now, since it's quite old and not ready for release. I'll reopen this once work starts on it again, but I don't want anyone mistakenly reviewing what I have here since it needs to be rebased and I'm not currently working on it.
